File: src/services/network-state.service.ts
import { Injectable, signal, computed } from '@angular/core';

const LOCAL_INFERENCE_KEY = 'pg_preferLocalInference';

@Injectable({ providedIn: 'root' })
export class NetworkStateService {
    private browserOnline = signal(typeof navigator !== 'undefined' ? navigator.onLine : true);
    readonly forceOffline = signal(false);

    /** When true, routes all AI calls to local PubGemma even when online. Persisted to localStorage. */
    readonly preferLocalInference = signal(
        typeof localStorage !== 'undefined'
            ? localStorage.getItem(LOCAL_INFERENCE_KEY) === 'true'
            : false
    );

    readonly isOnline = computed(() => this.browserOnline() && !this.forceOffline());

    /** Whether local inference should currently be used (offline OR user prefers it). */
    readonly useLocalInference = computed(() => !this.isOnline() || this.preferLocalInference());

    /**
     * Reactive label for the active AI provider shown in UI status tooltips.
     * Returns one of: 'Gemini Cloud', 'Local (PubGemma)', or 'Offline (PubGemma)'.
     */
    readonly activeProvider = computed(() => {
        if (!this.isOnline()) return 'Offline (PubGemma)';
        if (this.preferLocalInference()) return 'Local (PubGemma)';
        return 'Gemini Cloud';
    });

    constructor() {
        if (typeof window !== 'undefined') {
            window.addEventListener('online', () => this.browserOnline.set(true));
            window.addEventListener('offline', () => this.browserOnline.set(false));
        }
    }

    toggleForceOffline() {
        this.forceOffline.update(v => !v);
    }

    /**
     * Toggles the user's preference for local inference and persists the choice to localStorage.
     */
    togglePreferLocalInference() {
        const next = !this.preferLocalInference();
        this.preferLocalInference.set(next);
        if (typeof localStorage !== 'undefined') {
            localStorage.setItem(LOCAL_INFERENCE_KEY, String(next));
        }
    }
}

File: pocketgull_flutter/lib/services/network_state_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:connectivity_plus/connectivity_plus.dart';

/// Network state — tracks online/offline status and AI execution path.
///
/// Flutter parity with Angular `network-state.service.ts`.
/// Uses `connectivity_plus` for real network detection on mobile,
/// unlike the Angular version which relies on `navigator.onLine`.

const String _localInferenceKey = 'pg_preferLocalInference';

class NetworkStateModel {
  final bool isConnected;
  final bool forceOffline;
  final bool preferLocalInference;

  const NetworkStateModel({
    this.isConnected = true,
    this.forceOffline = false,
    this.preferLocalInference = false,
  });

  /// Effective online status accounting for forced offline toggle.
  bool get isOnline => isConnected && !forceOffline;

  /// Whether local inference should be used (offline OR user prefers it).
  bool get useLocalInference => !isOnline || preferLocalInference;

  /// Reactive label for the active AI provider shown in status UI.
  String get activeProvider {
    if (useLocalInference) {
      final prefix = isOnline ? 'Local' : 'Offline';
      return '$prefix (On-Device)';
    }
    return 'Gemini Cloud';
  }

  NetworkStateModel copyWith({
    bool? isConnected,
    bool? forceOffline,
    bool? preferLocalInference,
  }) {
    return NetworkStateModel(
      isConnected: isConnected ?? this.isConnected,
      forceOffline: forceOffline ?? this.forceOffline,
      preferLocalInference: preferLocalInference ?? this.preferLocalInference,
    );
  }
}

class NetworkStateNotifier extends Notifier<NetworkStateModel> {
  StreamSubscription<List<ConnectivityResult>>? _subscription;

  @override
  NetworkStateModel build() {
    _init();

    // Clean up subscription when provider is disposed.
    ref.onDispose(() {
      _subscription?.cancel();
    });

    return const NetworkStateModel();
  }

  Future<void> _init() async {
    // Load persisted preference.
    try {
      final prefs = await SharedPreferences.getInstance();
      final prefLocal = prefs.getString(_localInferenceKey) == 'true';
      if (prefLocal) {
        state = state.copyWith(preferLocalInference: true);
      }
    } catch (e) {
      debugPrint('[NetworkState] Failed to load preference: $e');
    }

    // Check current connectivity.
    try {
      final results = await Connectivity().checkConnectivity();
      final connected = !results.contains(ConnectivityResult.none);
      state = state.copyWith(isConnected: connected);
    } catch (e) {
      debugPrint('[NetworkState] Connectivity check failed: $e');
    }

    // Listen for changes.
    _subscription = Connectivity().onConnectivityChanged.listen((results) {
      final connected = !results.contains(ConnectivityResult.none);
      debugPrint('[NetworkState] Connectivity changed: $results → online=$connected');
      state = state.copyWith(isConnected: connected);
    });
  }

  /// Toggle forced offline mode for testing.
  void toggleForceOffline() {
    state = state.copyWith(forceOffline: !state.forceOffline);
    debugPrint('[NetworkState] Force offline: ${state.forceOffline}');
  }

  /// Toggle the user's preference for local inference and persist.
  Future<void> togglePreferLocalInference() async {
    final next = !state.preferLocalInference;
    state = state.copyWith(preferLocalInference: next);
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_localInferenceKey, next.toString());
    } catch (e) {
      debugPrint('[NetworkState] Failed to persist local inference pref: $e');
    }
  }
}

final networkStateProvider =
    NotifierProvider<NetworkStateNotifier, NetworkStateModel>(() {
  return NetworkStateNotifier();
});
